File: week-11/dahyen0o/표-편집.java
// package template;

import java.util.*;

class Solution {

    public String solution(final int n, int idx, String[] cmds) {
        final Stack<Integer> deleted = new Stack<>();

        final int[] prev = new int[n];
        final int[] next = new int[n];

        // init
        for (int i = 0; i < n; i++) {
            prev[i] = i - 1;
            next[i] = i + 1;
        }

        for (String cmd : cmds) {
            String[] c = cmd.split(" ");
            if ("U".equals(c[0])) {
                final int count = Integer.parseInt(c[1]);
                for (int cnt = 0; cnt < count; cnt++) {
                    idx = prev[idx];
                }
                continue;
            }
            if ("D".equals(c[0])) {
                final int count = Integer.parseInt(c[1]);
                for (int cnt = 0; cnt < count; cnt++) {
                    idx = next[idx];
                }
                continue;
            }
            if ("C".equals(c[0])) {
                deleted.push(idx);
                if (next[idx] >= n) { // 마지막 행 삭제
                    int nextIdx = prev[idx];
                    next[nextIdx] = n;
                    idx = nextIdx;
                    continue;
                }
                int nextIdx = next[idx];
                if (prev[idx] >= 0)
                    next[prev[idx]] = nextIdx;
                if (nextIdx < n)
                    prev[nextIdx] = prev[idx];
                idx = nextIdx;
                continue;
            }
            if ("Z".equals(c[0])) {
                final int restored = deleted.pop();
                // 복구할 자리로 이동
                if (prev[restored] >= 0)
                    next[prev[restored]] = restored;
                if (next[restored] < n)
                    prev[next[restored]] = restored;
                continue;
            }
            throw new RuntimeException("이상한 커맨드");
        }

        final boolean[] lived = new boolean[n];
        for (int d : deleted) {
            lived[d] = true;
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            if (!lived[i]) {
                sb.append("O");
                continue;
            }
            sb.append("X");
        }
        return sb.toString();
    }
}

File: week-11/bkryusim/표_편집.py
def solution(n, k, cmd):
    class Node:
        def __init__(self, prev, next, value):
            self.prev = prev
            self.next = next
            self.value = value

    answer = ['O'] * n
    stack = []
    head = Node(None, None, 0)

    prev = head
    for i in range(n):
        next = Node(prev, None, i)
        prev.next = next
        prev = next

    cursor = head
    for _ in range(k+1):
        cursor = cursor.next

    for line in cmd:
        command = line[0]
        if command == 'U':
            X = int(line.split()[1])
            for _ in range(X):
                cursor = cursor.prev

        if command == 'D':
            X = int(line.split()[1])
            for _ in range(X):
                cursor = cursor.next

        if command == 'C':
            stack.append(cursor)
            answer[cursor.value] = 'X'
            if cursor.prev:
                cursor.prev.next = cursor.next
            if cursor.next:
                cursor.next.prev = cursor.prev

            if cursor.next:
                cursor = cursor.next
            else:
                cursor = cursor.prev

        if command == 'Z':
            node = stack.pop()
            if node.prev:
                node.prev.next = node
            if node.next:
                node.next.prev = node
            answer[node.value] = 'O'
    return ''.join(answer)
